File: dataforge-fx/src/main/java/hep/dataforge/fx/values/ValueChooser.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.fx.values;

import hep.dataforge.description.ValueDescriptor;
import hep.dataforge.values.Value;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.value.ObservableObjectValue;
import javafx.scene.Node;

/**
 * A value chooser object. Must have an empty constructor to be invoked by
 * reflections.
 *
 * @author <a href="mailto:altavir@gmail.com">Alexander Nozik</a>
 */
public interface ValueChooser {

    //TODO replace by fx bindings
    void setValueCallback(ValueCallback callback);

    ValueCallback getValueCallback();

    ObjectProperty<Value> valueProperty();

    /**
     * The descriptor property for this value. Could be null
     *
     * @return
     */
    ObjectProperty<ValueDescriptor> descriptorProperty();

    /**
     * If showDefault, than default Value from descriptor is shown (if found)
     *
     * @return
     */
    BooleanProperty showDefaultProperty();

    /**
     * Set display value but do not notify listeners
     *
     * @param value
     */
    void setDisplayValue(Value value);

    /**
     * Set value and notify callback if it is present
     *
     * @param value
     */
    default void setValue(Value value) {
        valueProperty().set(value);
    }

    default void bindValue(ObservableObjectValue<Value> value) {
        valueProperty().bind(value);
    }

    default void setDescriptor(ValueDescriptor desc) {
        descriptorProperty().set(desc);
    }

    default void setShowDefault(boolean showDefault) {
        showDefaultProperty().set(showDefault);
    }

    default void setDisabled(boolean disabled) {
        //TODO replace by property
    }

    default ValueDescriptor getDescriptor() {
        return descriptorProperty().get();
    }

    /**
     * Get or create a Node that could be later inserted into some parent
     * object.
     *
     * @return
     */
    Node getNode();

}

File: dataforge-gui/src/main/kotlin/hep/dataforge/fx/values/ValueChooser.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.fx.values

import hep.dataforge.description.ValueDescriptor
import hep.dataforge.values.Value
import javafx.beans.property.ObjectProperty
import javafx.scene.Node

/**
 * A value chooser object. Must have an empty constructor to be invoked by
 * reflections.
 *
 * @author [Alexander Nozik](mailto:altavir@gmail.com)
 */
interface ValueChooser {

    /**
     * Get or create a Node that could be later inserted into some parent
     * object.
     *
     * @return
     */
    val node: Node

    /**
     * The descriptor property for this value. Could be null
     *
     * @return
     */
    val descriptorProperty: ObjectProperty<ValueDescriptor?>
    var descriptor: ValueDescriptor?

    val valueProperty: ObjectProperty<Value?>
    var value: Value?



    /**
     * Set display value but do not notify listeners
     *
     * @param value
     */
    fun setDisplayValue(value: Value)


    fun setDisabled(disabled: Boolean) {
        //TODO replace by property
    }

    fun setCallback(callback:ValueCallback)
}

object ValueChooserFactory{
    private fun build(descriptor: ValueDescriptor?): ValueChooser {
        if(descriptor == null){
            return TextValueChooser();
        }
        //val types = descriptor.type
        val chooser: ValueChooser = when {
            descriptor.allowedValues.isNotEmpty() -> ComboBoxValueChooser()
            descriptor.tags.contains("widget:color") -> ColorValueChooser()
            else -> TextValueChooser()
        }
        chooser.descriptor = descriptor
        return chooser
    }

    fun build(initialValue: Value, descriptor: ValueDescriptor? = null, callback: ValueCallback): ValueChooser {
        val chooser = build(descriptor)
        if (initialValue != Value.NULL) {
            chooser.setDisplayValue(initialValue)
        }
        chooser.setCallback(callback)
        return chooser
    }
}
